### ISSUE TITLE
Make `twisted.internet.reactor` import on Python 3

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5979]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)| @itamarst reported|
|-|-|
|Trac ID|trac#5979|
|Type|enhancement|
|Created|2012-09-13 20:46:36Z|
|Branch|https://github.com/twisted/twisted/tree/epollreactor-py3-5979-3|

Given the interleaved nature of `twisted.internet` and its tests, we probably can't port modules in isolation. Instead, we should first make the modules necessary for the basic reactor importable. Then when we port module by module, we can do so without having broken imports make it impossible.

<details><summary>Searchable metadata</summary>

```
trac-id__5979 5979
type__enhancement enhancement
reporter__itamar itamar
priority__normal normal
milestone__Python_3_3_Minimal Python 3.3 Minimal
branch__branches_epollreactor_py3_5979_3 branches/epollreactor-py3-5979-3
branch_author__itamarst itamarst
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1347569196000000 1347569196000000
changetime__1347984763000000 1347984763000000
version__None None
owner__itamar itamar

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_1"></a>|@itamarst commented|
|-|-|

(In [35687]) Branching to 'epollreactor-py3-5979'

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_2"></a>|@itamarst commented|
|-|-|

In order for this to be useful, we probably actually want `twisted.internet.reactor` to be importable, which slightly expands the scope of the ticket.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_3"></a>|@itamarst commented|
|-|-|

(In [35711]) Branching to 'epollreactor-py3-5979-2'

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_4"></a>|@itamarst set owner to @exarkun|
|-|-|

Ready for review. I tried to minimize necessary changes; probably they could've been slightly more minimal. `twisted.internet.reactor` imports now.

Started test run:
http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/epollreactor-py3-5979-2

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun set owner to @itamarst|
|-|-|

  1. Looks like the changes to abstract.py can be dropped now
  1. Not new in this branch, but `default.py` says "Linux: epoll(7) is the _fault_, since it scales well." which is correct in its way, but not what was intended I think.
  1. The udp and tcp imports now hidden inside reactor methods in posixbase.py should have comments to move back up to module scope once udp and tcp are ported
  1. The new `SyntaxError` for handling the unix import should have a similar comment, or unix should be handled the same way as udp and tcp (the latter is preferable, I think).
  1. Too bad about all of the extra changes.  It's pretty sad that `twisted.internet.task` is a dependency of twisted.internet.reactor.  That dependency seems to be due to `twisted.protocols.tls`?  Now I wish I'd worked a little harder on that module.  I want to ask for some ticket for fixing this, but I can't quite figure out what the ticket is.
  1. A comment about why there's a queue/Queue thing in threads.py would be nice
  1. Explaining what it means that things don't subclass Ephemeral anymore would probably be nice.  I bet most people have no clue what this class is or what value it provided.  Just an extra clause in the news fragment, like "; incorrectly pickling these types will no longer log a warning."
  1.  Some conflicts in _twistedpython3.py of course

Nothing too major here.  Merge when the above are addressed.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_6"></a>|@itamarst commented|
|-|-|

(In [35735]) Branching to 'epollreactor-py3-5979-3'

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_7"></a>|@itamarst commented|
|-|-|

1 and 8 were addressed by merge forward. 2, 3, 4, 6, 7 were addressed by 35737.

For `IReactorUnix`, I switched the check for whether to enable the interface to base it on the `platformType`, and moved imports in as with other tickets.

I opened ticket [#6004](https://github.com/twisted/twisted/issues/6004) to deal with issue 5.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_8"></a>|@itamarst commented|
|-|-|

Started build, will merge if it's happy: http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/epollreactor-py3-5979-3

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_9"></a>|@itamarst set status to `closed`|
|-|-|

(In [35740]) Merge epollreactor-py3-5979-3.

Author: itamar
Review: exarkun
Fixes: [#5979](https://github.com/twisted/twisted/issues/5979)

Make a bunch of modules import on Python 3, port twisted.internet.default and .reactor to Python 3, and remove Ephemeral base class from a bunch of reactor-related classes.
